system/lib/wasmfs/thread_utils.h

```diff
@@ -66,93 +66,158 @@ class SyncToAsync {
 
   static void* threadMain(void* arg) {
+    auto* parent = (SyncToAsync*)arg;
+
+    // The original thread should be waiting for us to start up. Acquire the
+    // lock to ensure the original thread has started waiting and to prevent it
+    // or any other thread from sending us work until we're ready.
+    parent->workerLock.lock();
+
+    // Now that we are safe from receiving work before we're ready, notify the
+    // original thread that it can continue. The original thread should be the
+    // only possible waiter because the SyncToAsync constructor has not returned
+    // yet.
+    parent->condition.notify_one();
+
+    // Schedule ourselves to start processing incoming work requests.
     emscripten_async_call(threadIter, arg, 0);
     return 0;
   }
 
+  // The main worker thread routine that waits for work, wakes up when work is
+  // available, executes the work, then schedules itself again.
   static void threadIter(void* arg) {
     auto* parent = (SyncToAsync*)arg;
-    if (parent->quit) {
-      pthread_exit(0);
-    }
+    assert(parent->workerLock.owns_lock());
+
     // Wait until we get something to do.
-    parent->childLock.lock();
-    parent->condition.wait(parent->childLock, [&]() {
-      return parent->readyToWork;
+    parent->state = Waiting;
+    parent->condition.wait(parent->workerLock, [&]() {
+      return parent->state == WorkAvailable || parent->state == ShouldExit;
     });
+
+    if (parent->state == ShouldExit) {
+      pthread_exit(0);
+    }
+
+    assert(parent->state == WorkAvailable);
     auto work = parent->work;
-    parent->readyToWork = false;
-    // Allocate a resume function, and stash it on the parent.
-    parent->resume = std::make_unique<std::function<void()>>([parent, arg]() {
-      // We are called, so the work was finished. Notify the caller.
-      parent->finishedWork = true;
-      parent->childLock.unlock();
-      parent->condition.notify_one();
+
+    // Allocate a resume function that will wake the invoker and schedule us to
+    // wait for more work.
+    parent->resume = [parent, arg]() {
+      assert(parent->workerLock.owns_lock());
+      // We are called, so the work was finished. Notify the invoker so it will
+      // wake up and continue once we resume waiting. There might be other
+      // invokers waiting to give us work, so `notify_all` to make sure our
+      // invoker wakes up.
+      parent->workCount++;
+      assert(parent->workCount && "workCount overflowed!");
+      parent->condition.notify_all();
+
       // Look for more work. Doing this asynchronously ensures that we continue
       // after the current call stack unwinds (avoiding constantly adding to the
       // stack, and also running any remaining code the caller had, like
       // destructors). TODO: add an option to do a synchronous call here in some
       // cases, which would avoid the time delay caused by a browser setTimeout.
       emscripten_async_call(threadIter, arg, 0);
```

Comment:
Holding locks during calls to `emscripten_async_call(threadIter, ...` is necessary to avoid missing work, but it also seems like a potentially large performance problem since it has to wait for `threadIter` to be rescheduled and called before the lock is released. Having an `invokeSync` method that does not rely on a callback for rescheduling could be more efficient because synchronous work could be performed in a loop without returning to the event loop.

Also, for the case of async work, it's unfortunate that multiple invoking threads cannot have async work in flight at once. This could be fixed by having a separate "state mailbox" for each invoking thread, but I'm not sure that's worth doing right now.